docs/rfc/rfc-018-bls-agg-exploration.md

```diff
@@ -0,0 +1,405 @@
+`G1` and `G2` which are points of the original curve _modulo_ different values.
+Finally, you define a third group `Gt`, where points from `G1` and `G2` satisfy
+the property of bilinearity with `Gt`. In this scheme, the function `e` takes
+as inputs points in `G1` and `G2` and outputs values in `Gt`. Succintly, given
+some point `P` in `G1` and some point `Q` in `G1`, `e(P, Q) = C` where `C` is in `Gt`.
+You can efficiently compute the mapping of points in `G1` and `G2` into `Gt`,
+but you cannot efficiently determine what points were summed and paired to
+produce the value in `Gt`.
+
+Functions are then defined to map digital signatures, messages, and keys into
+and out of points of `G1` or `G2` and signature verification is the process
+of calculating if a set of values representing a message, public key, and digital
+signature produce the same value in `Gt` through `e`.
+
+Signatures can be created as either points in `G1` with public keys being
+created as points in `G2` or vice versa. For the case of BLS12-381, the popular
+curve used, points in `G1` are represented with 48 bytes and points in `G2` are
+represented with 96 bytes. It is up to the implementer of the cryptosystem to
+decide which should be larger, the public keys or the signatures.
+
+BLS signatures rely on pairing-based elliptic-curve cryptography to produce
+various types of signatures. For a more in-depth but still high level discussion
+pairing-based elliptic-curve cryptography, see Vitalik Buterin's post on
+[Exploring Elliptic Curve Pairings][vitalik-pairing-post]. For much more in
+depth discussion, see the specific paper on BLS12-381, [Short signatures from
+ the Weil Pairing][bls-weil-pairing] and
+[Compact Multi-Signatures for Smaller Blockchains][multi-signatures-smaller-blockchains].
+
+### Adoption
+
+BLS signatures have already gained traction within several popular projects.
+
+* Algorand is working on an implementation.
+* [Zcash][zcash-adoption] has adopted BLS12-381 into the protocol.
+* [Ethereum 2.0][eth-2-adoption] has adopted BLS12-381 into the protocol.
+* [Chia Network][chia-adoption] has adopted BLS for signing blocks.
+* [Ostracon][line-ostracon-pr], a fork of Tendermint has adopted BLS for signing blocks.
+
+### What systems may be affected by adding aggregated signatures?
+
+#### Gossip
+
+Gossip could be updated to aggregate vote signatures during a consensus round.
+This appears to be of frankly little utility. Creating an aggregated signature
+incurs overhead, so frequently re-aggregating may incur a significant
+overhead. How costly this is is still subject to further investigation and
+performance testing.
+
+Even if vote signatures were aggregated before gossip, each validator would still
+need to receive and verify vote extension data from each (individual) peer validator in 
+order for consensus to proceed. That displaces any advantage gained by aggregating signatures across the vote message in the presence of vote extensions.
+
```

Comment:
I think that there are two main questions here, that should be considered independently.

First, the ability to aggregate votes should improve performance but the cost of aggregating signatures should definitely be considered in the equation. Notice that by aggregating single votes into a multi-vote messages, we don't only reduce the network overhead of sending multiple messages that are identical, but we also potentially speed up consensus. The problem is that, when considering that messages are signed, and are exactly signatures the only distinction between similar votes, the cost of aggregating signatures becomes a crucial element.

Putting this first point in a possibly simpler way. If vote messages were not signed, I think that no one would disagree that aggregating votes in the scale at which we run Tendermint would be clearly beneficial in terms of performance. However, messages are signed, and message signing and verification are probably the most expensive operations in this protocol. So, the cost of aggregating signatures + the cost of verifying aggregated signatures should be balanced with the existing costs of producing and verifying individual signatures. With today's technology and implementations, it appears that aggregated signatures are definitely effective when they can replace something like 100 individual signatures, and there is some discussion of the actual effect when replacing less, say around 40-60 signatures. But, below such thresholds, like mentioned in the text, this strategy does not pay off.

The second point, which is very important but somehow orthogonal, is about the design decision that was made of including additional, not redundant information on vote messages (e.g. vote extensions). I think that some of the appeal of this approach comes from the fact that we need "in any case" to propagate lots of vote messages, that are mostly identical, so why not use them for something else? This argument would be much less appealing if signature/vote aggregation was already adopted by Tendermint consensus.

Reply:
An additional information regarding the second point above. Notice that is possible to aggregate signatures on different messages, so adding non-redundant content to votes messages does not necessarily prevent the adoption of BLS signatures. Of course, this is a different user case, yet to be considered in terms of feasibility/efficiency.

Reply:
> However, messages are signed, and message signing and verification are probably the most expensive operations in this protocol.

I'm not totally convinced that this is true, at least _in the case of consensus_. In blocksync and light client verification, it may be true that the slow piece is signature verification but in consensus I'm less sure. Consensus involves sending a small packet of data to dozens or hundreds of networked nodes. Network transmission times in distributed networks are going to be in the dozens of milliseconds per message at least so it seems to me that the slow step in consensus is really transmitting vote information, not verifying signatures, which should be on the order of microseconds or nanoseconds per signature.

> I think that some of the appeal of this approach comes from the fact that we need "in any case" to propagate lots of vote messages, that are mostly identical, so why not use them for something else?

I would tend to agree. While the messages are called 'vote extensions' I do think we have a bit of freedom with their actually design and propagation. If we find a faster way to propagate all of the data that is not necessarily by putting them on the same message, that would certainly work.

> First, the ability to aggregate votes should improve performance but the cost of aggregating signatures should definitely be considered in the equation.

This is raising an important / interesting point. Signature and vote aggregation are somewhat orthogonal. Since, in Tendermint v0.36+, all validators sign over the same data, there isn't much need to retransmit all of the same information each time. We could certainly create combined vote messages that simultaneously transmit many signatures without including height/round/step information multiple times.